File: 77. SumRootToLeafNumbers.c
#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct TreeNode *newNode(int val) {
    struct TreeNode *node = malloc(sizeof(*node));
    node->val = val;
    node->left = node->right = NULL;
    return node;
}

int dfs(struct TreeNode *root, int curr) {
    if (!root) return 0;
    curr = curr * 10 + root->val;
    if (!root->left && !root->right) return curr;
    return dfs(root->left, curr) + dfs(root->right, curr);
}

int sumNumbers(struct TreeNode *root) {
    return dfs(root, 0);
}

int main(void) {
    /* Tree:
     *       4
     *      / \
     *     9   0
     *    / \
     *   5   1
     * Paths: 495 + 491 + 40 = 1026
     */
    struct TreeNode *root = newNode(4);
    root->left = newNode(9);
    root->right = newNode(0);
    root->left->left = newNode(5);
    root->left->right = newNode(1);

    printf("Sum: %d\n", sumNumbers(root));
    return 0;
}

File: 77. SumRootToLeafNumbers.java
public class sumRootToLeafNumbers {

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }
    }

    public static int sumNumbers(TreeNode root) {
        return dfs(root, 0);
    }

    private static int dfs(TreeNode node, int current) {
        if (node == null) {
            return 0;
        }

        int value = current * 10 + node.val;

        if (node.left == null && node.right == null) {
            return value;
        }

        return dfs(node.left, value) + dfs(node.right, value);
    }

    public static void main(String[] args) {
        //       4
        //      / \
        //     9   0
        //    / \
        //   5   1
        // Paths: 495 + 491 + 40 = 1026
        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(9);
        root.right = new TreeNode(0);
        root.left.left = new TreeNode(5);
        root.left.right = new TreeNode(1);

        System.out.println("Sum: " + sumNumbers(root));
    }
}
